src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -23,6 +26,98 @@ public static class Renamer
                 newName, optionSet, cancellationToken);
         }
 
+        /// <summary>
+        /// Similar to calling <see cref="Document.WithName(string)" /> with additional changes to the solution. 
+        /// Each change is added as a <see cref="RenameDocumentAction"/> in the returned <see cref="RenameDocumentActionSet.ApplicableActions" />.
+        /// 
+        /// Each action may individually encounter errors that prevent it from behaving correctly. Those are reported in <see cref="RenameDocumentAction.GetErrors(System.Globalization.CultureInfo?)"/>.
+        /// 
+        /// Current supported actions that may be returned: 
+        /// * <see cref="RenameSymbolDocumentAction"/> that will rename the type to match the document name
+        /// </summary>
+        public static async Task<RenameDocumentActionSet> RenameDocumentNameAsync(
+            Document document,
+            string newDocumentName,
+            OptionSet optionSet,
+            CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newDocumentName == null)
+            {
+                throw new ArgumentNullException(nameof(newDocumentName));
+            }
+
+            using var _ = ArrayBuilder<RenameDocumentAction>.GetInstance(out var actions);
+
+            if (!newDocumentName.Equals(document.Name))
```

Comment:
what does it mean for a client to ask us to try to rename a Document to its own name.  Should we fail in that case?

Reply:
this comment sitll applies.

Reply:
Personally I think it's fine to just "do nothing", or in this case, provide an empty action set. It's not really an error to rename a document to the same name so I don't want to throw an exception or fail. That said, this is the first public API I'm adding. If there's a reason we should fail or if it better fits the API design I'm happy to do that.